### ISSUE TITLE
 Bolt connection [/127.0.0.1:51026] will be closed because the client did not consume outgoing buffers for 00:15:00.000 which is not expected.

### ISSUE THREAD

**Issue by Meatballs1** [neo4j/neo4j#11997]
Server version 3.4.1
Client Bolt Driver .NET 1.6.1 

I was running a job that does post processing on my database. 

Memory settings (128Gb machine)

```
dbms.memory.heap.initial_size=32g
dbms.memory.heap.max_size=32g
dbms.memory.pagecache.size=62g
```

I got the following error after my program ran for about 30 minutes:

Server debug log:
```
2018-07-25 17:33:47.502+0000 ERROR [o.n.b.r.DefaultBoltConnection] Unexpected error detected in bolt session '0242acfffe110004-000019db-00000026-2abd7a414ed2a365-87e0e5e6'. Bolt connection [/127.0.0.1:51026] will be closed because the client did not consume outgoing buffers for 00:15:00.000 which is not expected.
org.neo4j.bolt.v1.runtime.BoltConnectionFatality: Bolt connection [/127.0.0.1:51026] will be closed because the client did not consume outgoing buffers for 00:15:00.000 which is not expected.
        at org.neo4j.bolt.v1.runtime.BoltStateMachine.handleFailure(BoltStateMachine.java:742)
        at org.neo4j.bolt.v1.runtime.BoltStateMachine.handleFailure(BoltStateMachine.java:728)
        at org.neo4j.bolt.v1.runtime.BoltStateMachine.handleFailure(BoltStateMachine.java:718)
        at org.neo4j.bolt.v1.runtime.BoltStateMachine.access$600(BoltStateMachine.java:62)
        at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$3.pullAll(BoltStateMachine.java:519)
        at org.neo4j.bolt.v1.runtime.BoltStateMachine.pullAll(BoltStateMachine.java:267)
        at org.neo4j.bolt.v1.messaging.BoltMessageRouter.lambda$onPullAll$6(BoltMessageRouter.java:114)
        at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:195)
        at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:143)
        at org.neo4j.bolt.runtime.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:170)
        at org.neo4j.bolt.runtime.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:153)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Client exception:

```
Neo4j.Driver.V1.ServiceUnavailableException: Connection with the server breaks due to AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer. ---> System.Net.Sockets.SocketException: Connection reset by peer
  at System.Net.Sockets.Socket.Receive (System.Byte[] buffer, System.Int32 offset, System.Int32 size, System.Net.Sockets.SocketFlags socketFlags) [0x00016] in <fc308f916aec4e4283e0c1d4b761760a>:0
  at System.Net.Sockets.NetworkStream.Read (System.Byte[] buffer, System.Int32 offset, System.Int32 size) [0x0009b] in <fc308f916aec4e4283e0c1d4b761760a>:0
   --- End of inner exception stack trace ---
  at System.Net.Sockets.NetworkStream.Read (System.Byte[] buffer, System.Int32 offset, System.Int32 size) [0x000e2] in <fc308f916aec4e4283e0c1d4b761760a>:0
  at Mono.Net.Security.MobileAuthenticatedStream+<>c__DisplayClass66_0.<InnerRead>b__0 () [0x0002b] in <fc308f916aec4e4283e0c1d4b761760a>:0
  at System.Threading.Tasks.Task`1[TResult].InnerInvoke () [0x0000f] in <71d8ad678db34313b7f718a414dfcb25>:0
  at System.Threading.Tasks.Task.Execute () [0x00010] in <71d8ad678db34313b7f718a414dfcb25>:0
--- End of stack trace from previous location where exception was thrown ---

```

When trying to reconnect to bolt I got the following errors and I needed to restart neo4j to connect:

```
Unhandled Exception:
Neo4j.Driver.V1.ClientException: Error when pulling unconsumed session.run records into memory in session: Connection with the server breaks due to IOException: Unexpected end of stream. ---> Neo4j.Driver.V1.ServiceUnavailableException: Connection with the server breaks due to IOException: Unexpected end of stream. ---> System.IO.IOException: Unexpected end of stream.
  at Neo4j.Driver.Internal.IO.ChunkReader.ReadNextMessages (System.IO.Stream messageStream) [0x00053] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.IO.BoltReader.Read (Neo4j.Driver.Internal.Messaging.IMessageResponseHandler responseHandler) [0x00001] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketClient.ReceiveOne (Neo4j.Driver.Internal.Messaging.IMessageResponseHandler responseHandler) [0x00047] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketClient.Receive (Neo4j.Driver.Internal.Messaging.IMessageResponseHandler responseHandler) [0x00004] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketConnection.Receive () [0x00016] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketConnection.Sync () [0x00008] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.DelegatedConnection.Sync () [0x00008] in <590164c3b042483288b8270a56ece478>:0
   --- End of inner exception stack trace ---
```

Don't have any more repro steps - I will monitor memory usage over another try to see if it works.

**Comment by Meatballs1** [comment]
If I roll back to 1.5.1 bolt driver I get:

```
Unhandled Exception:
Neo4j.Driver.V1.ServiceUnavailableException: Connection with the server breaks due to IOException: Unexpected end of stream. ---> System.IO.IOException: Unexpected end of stream.
  at Neo4j.Driver.Internal.IO.ChunkReader.ReadNextMessages (System.IO.Stream messageStream) [0x00053] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.IO.BoltReader.Read (Neo4j.Driver.Internal.Messaging.IMessageResponseHandler responseHandler) [0x00001] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketClient.ReceiveOne (Neo4j.Driver.Internal.Messaging.IMessageResponseHandler responseHandler) [0x00047] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.SocketConnection.ReceiveOne () [0x00001] in <590164c3b042483288b8270a56ece478>:0
  at Neo4j.Driver.Internal.Connector.DelegatedConnection.ReceiveOne () [0x00008] in <590164c3b042483288b8270a56ece478>:0
   --- End of inner exception stack trace ---

```

but no error in neo4j debug.log

**Comment by Meatballs1** [comment]
I tried a slightly simpler query and received the following after a short while:

```

Unhandled Exception:
Neo4j.Driver.V1.ServiceUnavailableException: Connection with the server breaks due to AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.IO.IOException: Remote prematurely closed connection.
  at Mono.Net.Security.AsyncProtocolRequest+<ProcessOperation>d__24.MoveNext () [0x000e7] in <fc308f916aec4e4283e0c1d4b761760a>:0
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <71d8ad678db34313b7f718a414dfcb25>:0
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <71d8ad678db34313b7f718a414dfcb25>:0
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <71d8ad678db34313b7f718a414dfcb25>:0
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <71d8ad678db34313b7f718a414dfcb25>:0
  at Mono.Net.Security.AsyncProtocolRequest+<StartOperation>d__23.MoveNext () [0x0008b] in <fc308f916aec4e4283e0c1d4b761760a>:0
```

**Comment by Meatballs1** [comment]
3.4.4 & 1.6.1 seems to work fine though \o/

**Comment by ali-ince** [comment]
Hi @Meatballs1,

The error message `Bolt connection [/127.0.0.1:51026] will be closed because the client did not consume outgoing buffers for 00:15:00.000 which is not expected.
org.neo4j.bolt.v1.runtime.BoltConnectionFatality: Bolt connection [/127.0.0.1:51026] will be closed because the client did not consume outgoing buffers for 00:15:00.000 which is not expected.` states that the server stopped generating data for 15 minutes (which is because the client did not read pending buffers) and terminates the connection since this is not an expected use case.

Could you please verify that your client code consumes results as they're generated by the server?

Thanks.

**Comment by Meatballs1** [comment]
@ali-ince Could it be that the neo4j server is taking 15 mins or longer to complete the query? I thought I'd seen some queries take longer than this (fairly complex query).

Anyway the actual client doesn't really care about the response - the query is used to compute a number of new edges and finishes with:

```
                                       FOREACH (x in isuser|
                                           MERGE (gpo)-[:LINK{cost:{cost}}]->(allowedNode)
                                        )
```

and is just run with:

```
session.WriteTransaction(tx => tx.Run(query, parameters));
```

I'm not bothered about the return value but would have thought the bolt client still would perform any necessary communication. Should I get and consume the return value?

**Comment by Meatballs1** [comment]
```
                    IStatementResult result = session.WriteTransaction(tx => tx.Run(query, parameters));
                    result.Consume();
```
?

**Comment by Meatballs1** [comment]
Definitely can run a 15+ min query without calling Consume() on the result without it failing. However same query/different parameters sometime later gets this error

**Comment by Meatballs1** [comment]
Calling consume didn't appear to make any difference. I think it is probably something getting stuck server side tbh

**Comment by Meatballs1** [comment]
The code in https://github.com/neo4j/neo4j/blob/97a961df92eae192fc1579d4995b81d69168b26a/community/bolt/src/main/java/org/neo4j/bolt/transport/TransportWriteThrottle.java#L77 is waiting for channel.isWritable(), which is going to depend on if we have fallen back past the low watermark (after hitting the high watermark). 

It seems like we aren't removing stuff from the queue - but at this point my client is only running in a single thread, doing a single call, and waiting for the server to return the result? So is the bolt client not consuming some kind of keep alive?

**Comment by Meatballs1** [comment]
I figured I would just catch the exception and retry, which made me realise that it wasn't where I thought it was!

* I have a query which returns a few hundred items from the DB
* I iterate over each of the item and perform a long running query on each item - sometimes this can be 30+ minutes to process each query.
* The outer query doesn't time out on first query that takes longer than 15 minutes - I can get through a number of queries - some that 30 mins+ and then a number less than a minute. 

```
                    Driver d;
                    Session session = d.Session();
                    IStatementResult result = session.Run(query);
                    foreach (var record in result)
                    {
                        String heavyQuery = "Query that does lots of CPU usage and can run for a long time";
                         Session innerSession = d.Session();
                         innerSession.Run(heavyQuery);
                    }
```

The exception is raised from the outer `session.Run()`

**Comment by ali-ince** [comment]
Hi @Meatballs1,

Your detailed use case and what you have been experiencing makes sense. 

It would be better if you get the first result into a list and iterate on this list rather than the `IStatementResult`. This way you will not run into this issue (because server will complete the query and will not try to send data to the client - which is busy executing queries in loop).

If your first query takes too much time to complete and you're trying to avoid the waiting time to start your loop, you should consider using a `BlockingCollection` as demonstrated [here](https://docs.microsoft.com/en-us/dotnet/standard/collections/thread-safe/blockingcollection-overview).

Thanks.

**Comment by Meatballs1** [comment]
@ali-ince Cool thanks,  I've gone for the List approach already and everything seems to be working. I'll keep in mind the blocking approach :D 

I'm a bit confused why a long running query of 30+ mins doesn't cause the 15 minute timeout initially, but it maybe because we haven't yet triggered the highwatermark condition?

**Comment by ali-ince** [comment]
@Meatballs1 The server probably sends more than one record in each round-trip, so you already have some items buffered in the statement result and for the error to be propagated you need to get to a position where client asks for more data (after buffered ones are processed in this case).

**Comment by Meatballs1** [comment]
Great, will close it then :)

**Comment by ecdavis** [comment]
I recently observed the same behavior after upgrading from neo4j 3.2.* to 3.5.* (yes, a big step). The Python driver doesn't handle this too gracefully -- I opened https://github.com/neo4j/neo4j-python-driver/issues/301 to track the issue I saw there.

Interestingly in our case the results of the queries aren't used, but so far the only way we've found to resolve this is by specifically consuming the results.
